src/Tools/BuildValidator/CompilationDiff.cs

```diff
@@ -42,41 +51,89 @@ internal sealed class CompilationDiff
             string ILFilePath,
             string CustomDataFilePath);
 
-        public bool? AreEqual { get; }
-        public string OriginalPath { get; }
-        public ImmutableArray<Diagnostic> Diagnostics { get; }
+        private readonly ImmutableArray<Diagnostic> _diagnostics;
+        private readonly byte[]? _originalPortableExecutableBytes;
+        private readonly byte[]? _rebuildPortableExecutableBytes;
+        private readonly Compilation? _rebuildCompilation;
+        private readonly ImmutableArray<MetadataReferenceInfo> _references;
+        private readonly LocalReferenceResolver? _localReferenceResolver;
+        private readonly string? _message;
+
+        public AssemblyInfo AssemblyInfo { get; }
+        public CompilationOutcome Outcome { get; }
+
+        public bool Succeeded => Outcome == CompilationOutcome.Success;
 
-        private CompilationDiff(string originalPath, bool? areEqual)
+        public ImmutableArray<Diagnostic> Diagnostics
         {
-            AreEqual = areEqual;
-            OriginalPath = originalPath;
+            get
+            {
+                Debug.Assert(Outcome == CompilationOutcome.CompilationError);
+                return _diagnostics;
+            }
         }
 
-        private CompilationDiff(ImmutableArray<Diagnostic> diagnostics, string originalPath)
+        public string MiscErrorMessage
         {
-            Diagnostics = diagnostics;
-            OriginalPath = originalPath;
+            get
+            {
+                Debug.Assert(Outcome == CompilationOutcome.MiscError);
```

Comment:
>Debug.Assert(Outcome == CompilationOutcome.MiscError); [](start = 16, length = 54)

Why require a specific `CompilationOutcome`?

Reply:
This is an ad-hoc discriminated union hence this property is only valid in this state.